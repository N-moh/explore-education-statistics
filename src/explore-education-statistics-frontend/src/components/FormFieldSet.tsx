import classNames from 'classnames';
import React, { FunctionComponent, ReactNode } from 'react';

export interface FieldSetProps {
  legend: string | ReactNode;
  legendSize?: 'xl' | 'l' | 'm' | 's';
  hint?: string;
  hintId?: string;
}

let idCounter = 0;

const FormFieldSet: FunctionComponent<FieldSetProps> = ({
  children,
  hint,
  hintId = `formFieldSetHint-${(idCounter += 1)}`,
  legend,
  legendSize = 'm',
}) => {
  return (
    <fieldset
      className="govuk-fieldset"
      aria-describedby={hint ? hintId : undefined}
    >
      <legend
        className={classNames(
          'govuk-fieldset__legend',
          `govuk-fieldset__legend--${legendSize}`,
        )}
      >
        {legend}
      </legend>
      {hint && (
        <span className="govuk-hint" id={hintId}>
          {hint}
        </span>
      )}
      {children}
    </fieldset>
  );
};

export default FormFieldSet;
