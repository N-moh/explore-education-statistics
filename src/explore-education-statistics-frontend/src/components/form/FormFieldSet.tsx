import classNames from 'classnames';
import React, { FunctionComponent, ReactNode } from 'react';
import ErrorMessage from '../ErrorMessage';
import FormGroup from './FormGroup';
import createDescribedBy from './util/createDescribedBy';

export interface FieldSetProps {
  error?: string;
  hint?: string;
  id: string;
  legend?: ReactNode | string;
  legendSize?: 'xl' | 'l' | 'm' | 's';
}

const FormFieldSet: FunctionComponent<FieldSetProps> = ({
  children,
  error,
  hint,
  id,
  legend,
  legendSize = 'm',
}) => {
  return (
    <FormGroup hasError={!!error}>
      <fieldset
        className="govuk-fieldset"
        id={id}
        aria-describedby={createDescribedBy({
          id,
          error: !!error,
          hint: !!hint,
        })}
      >
        {legend && (
          <legend
            className={classNames(
              'govuk-fieldset__legend',
              `govuk-fieldset__legend--${legendSize}`,
            )}
          >
            {legend}
          </legend>
        )}
        {hint && (
          <span className="govuk-hint" id={`${id}-hint`}>
            {hint}
          </span>
        )}
        {error && <ErrorMessage id={`${id}-error`}>{error}</ErrorMessage>}
        {children}
      </fieldset>
    </FormGroup>
  );
};

export default FormFieldSet;
