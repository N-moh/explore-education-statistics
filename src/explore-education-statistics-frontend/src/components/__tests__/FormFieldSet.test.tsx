import React from 'react';
import { render } from 'react-testing-library';
import FormFieldSet from '../FormFieldSet';

describe('FormFieldSet', () => {
  test('renders correctly with required props', () => {
    const { container } = render(
      <FormFieldSet legend="Fill the form">
        <input type="text" name="test-input-1" />
        <input type="text" name="test-input-2" />
      </FormFieldSet>,
    );

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders correctly with hint', () => {
    const { container, getByText } = render(
      <FormFieldSet
        legend="Fill the form"
        hint="All fields are required"
        hintId="the-hint"
      >
        <input type="text" name="test-input-1" />
        <input type="text" name="test-input-2" />
      </FormFieldSet>,
    );

    expect(getByText('All fields are required')).toBeDefined();
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('auto-generates `hintId` when none is set', () => {
    const { getByText } = render(
      <FormFieldSet legend="Fill the form" hint="All fields are required">
        <input type="text" name="test-input-1" />
        <input type="text" name="test-input-2" />
      </FormFieldSet>,
    );

    expect(getByText('All fields are required').id).toContain(
      'formFieldSetHint',
    );
  });

  test('setting `legendSize` to xl applies class correctly', () => {
    const { container } = render(
      <FormFieldSet
        legend="Fill the form"
        hint="All fields are required"
        legendSize="xl"
      >
        <input type="text" name="test-input-1" />
        <input type="text" name="test-input-2" />
      </FormFieldSet>,
    );

    expect(container.querySelector('legend')).toHaveClass(
      'govuk-fieldset__legend--xl',
    );
  });
});
