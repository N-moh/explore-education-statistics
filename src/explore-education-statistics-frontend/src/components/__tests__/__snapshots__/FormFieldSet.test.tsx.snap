// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFieldSet renders correctly with hint 1`] = `

<fieldset class="govuk-fieldset"
          aria-describedby="the-hint"
>
  <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">
    Fill the form
  </legend>
  <span class="govuk-hint"
        id="the-hint"
  >
    All fields are required
  </span>
  <input type="text"
         name="test-input-1"
  >
  <input type="text"
         name="test-input-2"
  >
</fieldset>

`;

exports[`FormFieldSet renders correctly with required props 1`] = `

<fieldset class="govuk-fieldset">
  <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">
    Fill the form
  </legend>
  <input type="text"
         name="test-input-1"
  >
  <input type="text"
         name="test-input-2"
  >
</fieldset>

`;
